xbmc/interfaces/json-rpc/AudioLibrary.cpp

```diff
@@ -51,16 +51,28 @@ JSONRPC_STATUS CAudioLibrary::GetArtists(const std::string &method, ITransportLa
   if (!musicUrl.FromString("musicdb://artists/"))
     return InternalError;
 
+  bool allroles = false;
+  if (parameterObject["allroles"].isBoolean())
+    allroles = parameterObject["allroles"].asBoolean();
+  
   int genreID = -1, albumID = -1, songID = -1;
   const CVariant &filter = parameterObject["filter"];
-  if (filter.isMember("genreid"))
-    genreID = (int)filter["genreid"].asInteger();
-  else if (filter.isMember("genre"))
-    musicUrl.AddOption("genre", filter["genre"].asString());
+  
+  if (allroles)
+    musicUrl.AddOption("roleid", -1000); //All roles, any negative parameter overrides implicit roleid=1 filter required for backward compatibility
   else if (filter.isMember("roleid"))
     musicUrl.AddOption("roleid", (int)filter["roleid"].asInteger());
   else if (filter.isMember("role"))
     musicUrl.AddOption("role", filter["role"].asString());
+  // Only one of (song) genreid/genre, albumid/album or songid/song or rules type filter is allowed by filter syntax
+  if (filter.isMember("genreid"))  //Deprecated. Use "songgenre" or "artistgenre"
+    genreID = (int)filter["genreid"].asInteger();
+  else if (filter.isMember("genre"))
+    musicUrl.AddOption("genre", filter["genre"].asString());
+  if (filter.isMember("songgenreid"))
```

Comment:
Yes but previously all was a big else if block.

Since now it's multiple block the if are hidden in the middle of the else if and this make the code not readable. And error prone in the future if new fields added.
At least add empty lines to separate the block and avoid the confusion.

The remark also covers that if you add the 2 fields in JSON query AddOption will be called twice, leading from just this code reading to unexpected behavior. 

And allowing both fields to actually do something is against your comment just before that says only 1 allowed.